src/server/pipeline.py

```diff
@@ -209,8 +214,18 @@ async def perform_handshake(
             raise Exception("No longer accepting handshakes, closing.")
 
         if not global_connections.add(connection):
+            # Don't penalize a duplicate connection.
+            mark_attempted = False
             raise ProtocolError(Err.DUPLICATE_CONNECTION, [False])
 
+        if (
+            not connection.is_outbound
+            and connection.connection_type == NodeType.FULL_NODE
```

Comment:
Need to also prevent other connection types. Perhaps we can have a whitelist for usecases where we are over the limit, but still want to accept 1 more connection.